replica/table.cc

```diff
@@ -666,11 +672,11 @@ class tablet_storage_group_manager final : public storage_group_manager {
         auto [id, side] = tablet_map().get_tablet_id_and_range_side(t);
         auto idx = id.value();
 #ifndef SCYLLA_BUILD_MODE_RELEASE
-        if (idx >= storage_groups().size()) {
+        if (idx >= tablet_count()) {
             on_fatal_internal_error(tlogger, format("storage_group_of: index out of range: idx={} size_log2={} size={} token={}",
-                                                    idx, log2_storage_groups(), storage_groups().size(), t));
+                                                    idx, log2_storage_groups(), tablet_count(), t));
         }
-        auto& sg = storage_groups()[idx];
+        auto& sg = storage_groups().at(idx);
```

Comment:
this is not backward compatible. I'm not sure if this function can be called for non-existent storage groups or not, but if it can, we should either catch and ignore the out_of_range exception here, or check if storage_groups().contains(idx) first for the sanity check.

Reply:
How can it be called for non-existent storage groups?

Reply:
From `storage_group_for_token`.
Even if we end up in requiring the storage_group's existence on this path, it should be done in a follow-up series, and leave this change backward compatible with the existing code, otherwise, we'll be changing too many things at once.

Reply:
This code happens to work now, but it will be fragile with dynamic allocation of storage_groups, unless we make sure we deal with non-existing storage_groups correctly.

Reply:
This function returns the index of the storage_group that holds a token, not a reference to the  storage_group.
There is no documentation saying the corresponding storage_group must be allocated on this shard.
We can change the semantics if we want to, but I still think we should do that in 2 steps so we can review this PR more easily and make it sure it doesn't break anything before applying the next step[, and second - we have to document the semantics of these functions, and/or have a self-explanatory api, like returning a storage_group& here rather than size_t.

Reply:
Only for the sanity check, and we don't know if it breaks with dynamic storage_group allocation.
@Deexie ?

Reply:
I am adding the sanity check as you proposed, will send new version and we can continue from there.